fairscale/experimental/nn/data_parallel/gossip/distributed.py

```diff
@@ -0,0 +1,1192 @@
+        )
+
+        self.old_params = torch.empty(self.world_portion_length, dtype=params_dtype).to(params_device).detach()
+
+        # copy params to old_params to initialize old_params
+        offset = 0
+        for group in optimizer.param_groups:
+            for p in group["params"]:
+                numel = p.numel()
+
+                if offset + numel > self.portion_start and offset < self.portion_end:
+
+                    # start and end for each
+                    overall_start = max(self.portion_start, offset)
+                    overall_end = min(self.portion_end, offset + numel)
+
+                    p_start = overall_start - offset
+                    p_end = overall_end - offset
+
+                    buffer_start = overall_start - self.portion_start
+                    buffer_end = overall_end - self.portion_start
+
+                    # let's see size of p and split based on that
+                    current_p = p.view(-1)[p_start:p_end]
+                    current_p_old = self.old_params[buffer_start:buffer_end]
+
+                    current_p_old.copy_(current_p)
+
+                offset += numel
+
+        self.global_momentum_buffer = torch.zeros_like(self.old_params).detach()
+
+    def _distributed_comm(self, optimizer: torch.optim.Optimizer, mode: str) -> None:
+        """ Performs the communication needed for the efficient SlowMo implementation """
+        offset = 0
+        slowmo_comm_lists: List[List[torch.Tensor]] = [[] for _ in range(self.slowmo_num_shards)]
+        with torch.no_grad():
+            for group in optimizer.param_groups:
+                # aggregate different parts of p in required node
+                for p in group["params"]:
+                    numel = p.numel()
+
+                    # gather has a reduce operation so division by world size is needed
+                    if mode == "gather":
+                        p /= self.process_world_size
+
+                    current_start = offset
+                    while current_start < offset + numel:
+                        main_node = current_start // self.world_portion_length
+
+                        main_node_end = (main_node + 1) * self.world_portion_length
+                        current_end = min(offset + numel, main_node_end)
+
+                        p_start = current_start - offset
+                        p_end = current_end - offset
+
+                        slowmo_comm_lists[main_node].append(p.view(-1)[p_start:p_end])
+
+                        current_start = current_end
+                    offset += numel
+
+            for slowmo_rank, slowmo_comm_list in enumerate(slowmo_comm_lists):
+                if mode == "gather":
+                    communication_op = functools.partial(dist.reduce, dst=slowmo_rank)
+                elif mode == "scatter":
+                    communication_op = functools.partial(dist.broadcast, src=slowmo_rank)
+                communicate(slowmo_comm_list, communication_op)
```

Comment:
it looks like something which could be sped up by using flat buffers + tensor views behind the scene (OSS/SDP/FSDP do that), instead of flattening everytime you communicate, unless I'm missing something ? There are some [helpers](https://github.com/facebookresearch/fairscale/blob/f2af4c66c70c0f673ddc7532b8c3db6954d67706/fairscale/nn/misc/param_bucket.py#L38) for that

Reply:
Thanks for the comment! I looked into the code for the different Buckets.

The reason we keep flattening and unflattening was to save memory from being used up. To explain a bit more, the communicate(...) happens after optimizer.step() - so its a time when not much memory is being used elsewhere (gradients, activations etc). This allows us to do the flattening and communication without worrying about the memory overhead.

It seems to me that you might be suggesting to keep flat buffers in a persistent manner throughout training. This looks like it might lead to larger memory usage as compared with flattening and unflattening as needed. I am not sure the tradeoff of speed will be worth the increase in memory usage.

Do let me know in case I might be missing something